File: opencode/src/events.ts
/**
 * Emit AgentEvent frames via the engine's stream builtin. Same per-process
 * epoch + per-session monotonic sequence scheme as the harness emitter, so
 * item_ids never collide across restarts.
 */

import { randomUUID } from 'node:crypto';
import type { IIIClient } from 'iii-sdk';
const PROCESS_EPOCH = randomUUID();
const seqBySession = new Map<string, number>();

export function makeEmitter(iii: IIIClient, streamName: string) {
  return async function emit(session_id: string, event: unknown): Promise<void> {
    const seq = seqBySession.get(session_id) ?? 0;
    seqBySession.set(session_id, seq + 1);
    const item_id = `${session_id}-${PROCESS_EPOCH}-${seq.toString().padStart(8, '0')}`;
    try {
      await iii.trigger({
        function_id: 'stream::set',
        payload: { stream_name: streamName, group_id: session_id, item_id, data: event },
      });
    } catch (err) {
      console.warn(`stream::set failed for ${session_id}: ${String(err)}`);
    }
  };
}

export type Emit = ReturnType<typeof makeEmitter>;

File: codex/src/events.rs
//! Stream emitter: writes frames via the engine's `stream::set` builtin with a
//! per-process epoch + per-session monotonic sequence so item_ids never
//! collide across restarts (matches the harness emitter). Failures are logged,
//! not propagated — the streams are best-effort observability; the function's
//! return value and the session record are the source of truth.

use std::collections::HashMap;
use std::sync::Mutex;

use iii_sdk::protocol::TriggerRequest;
use iii_sdk::IIIClient;
use once_cell::sync::Lazy;
use serde_json::{json, Value};
use uuid::Uuid;

static EPOCH: Lazy<String> = Lazy::new(|| Uuid::new_v4().to_string());
static SEQ: Lazy<Mutex<HashMap<String, u64>>> = Lazy::new(|| Mutex::new(HashMap::new()));

fn next_item_id(session_id: &str) -> String {
    let mut map = SEQ.lock().expect("seq mutex");
    let n = map.entry(session_id.to_string()).or_insert(0);
    let seq = *n;
    *n += 1;
    format!("{session_id}-{}-{seq:08}", *EPOCH)
}

pub async fn emit(iii: &IIIClient, stream_name: &str, session_id: &str, data: Value) {
    let item_id = next_item_id(session_id);
    let res = iii
        .trigger(TriggerRequest {
            function_id: "stream::set".to_string(),
            payload: json!({
                "stream_name": stream_name,
                "group_id": session_id,
                "item_id": item_id,
                "data": data,
            }),
            action: None,
            timeout_ms: Some(5_000),
        })
        .await;
    if let Err(e) = res {
        tracing::warn!(stream_name, session_id, error = %e, "stream::set failed");
    }
}
